File: greedy/452.Minimum_Number_of_Arrows_to_Burst_Balloons.py
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:

        # Similar to 435. Non-overlapping Intervals
        # Greedy
        # find the # of non-overlapping intervals
        if len(points) <= 1:
            return len(points)
        non_overlapping_count = 1  # inclusive
        points.sort(key=lambda x : x[1])
        prev_end = points[0][1]
        for i in range(1, len(points)):
            curr_start = points[i][0]
            if curr_start > prev_end:
                non_overlapping_count += 1
                prev_end = points[i][1]
        
        return non_overlapping_count

File: greedy/452.Minimum_Number_of_Arrows_to_Burst_Balloons.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int findMinArrowShots(vector<vector<int>>& points) {
        
        if (points.size() < 2)
            // return static_cast<int>(points.size());
            return points.size();
        
        sort(points.begin(), points.end(), [](vector<int>& a, vector<int>& b){
            return a[1] < b[1];
        });
        int min_arrows_required = 1; 
        int prev_end = points[0][1];
        for (int i = 1; i < points.size(); ++i)
        {
            if (points[i][0] > prev_end)
            {
                ++min_arrows_required;
                prev_end = points[i][1];
            }
        }
        return min_arrows_required;
    }
};
